candidatebot.py

```diff
@@ -17,9 +18,44 @@
 XML_FILE = "CandidateSummaryAction.xml"
 # Limit what this does during testing.
 MAX_PAGES_TO_CREATE = 3
+# Rate-limit to five page-creations every second. Change to 0.1 before running
+# against Wikipedia, but it's fine to hammer cso.noidea.dog.
+EDIT_PAGES_PER_SECOND = 4
+QUERY_PAGES_PER_SECOND = 20
 
 # TODO: Set a user agent.
 
+def rate_limited(max_per_second):
+  """Rate limiting decorator-with-args.
+  Args:
+    max_per_second: (float) How many times per second to do the thing.
+  Returns:
+    (func): A decorator
+  """
+  interval = 1.0 / float(max_per_second)
+
+  def decorator(func):
+    """A rate-limiting decorator.
+
+    Args:
+      func: (func) The thing to wrap.
+    Returns:
+      (func): The rate-limting function to apply to the thing to wrap.
+    """
+    last_called = [0.0]
+
+    def rate_limited_function(*args, **kwargs):
+      """The actual rate limiting logic."""
+      elapsed = time.time() - last_called[0]
+      wait = interval - elapsed
+      if wait > 0:
+        time.sleep(wait)
+      ret = func(*args, **kwargs)
+      last_called[0] = time.time()
```

Comment:
It can't, because it'll be constant inside the closure and get reset every time. But the list doesn't get reset, so this is how we keep track of the 'last_called' time between calls.